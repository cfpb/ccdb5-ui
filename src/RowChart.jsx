import './RowChart.less'
import * as d3 from 'd3'
import { connect } from 'react-redux'
import { max } from 'd3-array'
import React from 'react'
import { row } from 'britecharts'

export class RowChart extends React.Component {
  constructor( props ) {
    super( props )
    this.aggtype = props.aggtype
  }

  _getHeight( numRows ) {
    return numRows === 1 ? 100 : numRows * 60
  }

  _wrapText( text, width ) {
    // ignore test coverage since this is code borrowed from d3 mbostock
    // text wrapping functions
    /* istanbul ignore next */
    // eslint-disable-next-line complexity
    text.each( function() {
      const innerText = d3.select( this )
      if ( innerText.node().children.length > 0 ) {
        // assuming its already split up
        return
      }
      const words = innerText.text().split( /\s+/ ).reverse(),
            // ems
            lineHeight = 1.1,
            y = innerText.attr( 'y' ) || 0,
            dy = parseFloat( innerText.attr( 'dy' ) || 0 )

      let word,
          line = [],
          lineNumber = 0,
          tspan = innerText.text( null ).append( 'tspan' ).attr( 'x', 0 )
            .attr( 'y', y ).attr( 'dy', dy + 'em' )

      // eslint-disable-next-line no-cond-assign
      while ( word = words.pop() ) {
        line.push( word )
        tspan.text( line.join( ' ' ) )
        if ( tspan.node().getComputedTextLength() > width ) {
          line.pop()
          tspan.text( line.join( ' ' ) )
          line = [ word ]
          tspan = innerText.append( 'tspan' )
            .attr( 'x', 0 )
            .attr( 'y', y )
            // eslint-disable-next-line no-mixed-operators
            .attr( 'dy', ++lineNumber * lineHeight + dy + 'em' )
            .text( word )
        }
      }
    } )
  }

  componentDidUpdate() {
    const data = this.props.data
    if ( !data || !data.length ) {
      return
    }

    this._redrawChart( data )
  }

  // --------------------------------------------------------------------------
  // Event Handlers
  _redrawChart( data ) {
    const rowData = data.slice( 0, 5 )
    const total = this.props.total
    const ratio = total / max( rowData, o => o.value )
    const chartID = '#row-chart-' + this.aggtype
    d3.select( chartID + ' .row-chart' ).remove()
    const rowContainer = d3.select( chartID )
    const width = rowContainer.node().getBoundingClientRect().width
    const height = this._getHeight( rowData.length )
    const chart = row()
    const marginLeft = width / 3
    chart.margin( {
      left: marginLeft,
      right: 0,
      top: 20,
      bottom: 10
    } )
      .backgroundColor( '#f7f8f9' )
      .enableLabels( true )
      .labelsSize( 14 )
      .labelsTotalCount( total.toLocaleString() )
      .labelsNumberFormat( ',d' )
      .outerPadding( 0.1 )
      .percentageAxisToMaxRatio( ratio )
      .yAxisLineWrapLimit( 1 )
      .yAxisPaddingBetweenChart( 5 )
      .width( width )
      .wrapLabels( false )
      .height( height )
    rowContainer.datum( rowData ).call( chart )
    this._wrapText( d3.select( chartID ).selectAll( '.tick text' ), marginLeft )
  }

  render() {
    return (
      <div>
        <h3>{ this.aggtype } by complaint volume (5 highest counts)</h3>
        <div id={ 'row-chart-' + this.aggtype }>
        </div>
      </div>
    )
  }
}

export const mapStateToProps = ( state, ownProps ) => {
  // use state.query to rip filter out the bars
  const aggtype = ownProps.aggtype
  const filters = state.query[aggtype]
  let data = state.map[aggtype]

  if ( filters && filters.length ) {
    data = data.filter( o => filters.includes( o.name ) )
  }
  return {
    data,
    total: state.results.total
  }
}

export default connect( mapStateToProps )( RowChart )
