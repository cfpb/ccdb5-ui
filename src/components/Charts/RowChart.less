.row-chart-section {
  margin-top: @grid_gutter-width;
  margin-left: @grid_gutter-width;

  p {
    max-width: none;
  }
}

@media @phone {
  .row-chart-section {
    margin-left: 0;
  }
}

.row-chart {
  .bg-hover,
  .visibility,
  line,
  path.domain,
  .x-axis-group {
    display: none;
  }

  .expandable {
    color: @pacific;
  }

  text {
    font-size: 16px;

    &.adjust-upwards {
      transform: translateY(-12px);
    }

    &.hidden {
      display: none;
    }
  }

  .view-more-group {
    .view-more-background {
      fill: @pacific-20;
    }

    .view-more-label {
      font-size: 14px;
      font-weight: bold;
      text-decoration: underline;
      fill: @pacific;
      text-anchor: end;
    }
  }
}
