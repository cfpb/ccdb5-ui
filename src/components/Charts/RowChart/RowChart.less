.row-chart-section {
  margin-top: @grid_gutter-width;
  margin-left: @grid_gutter-width;

  p {
    max-width: none;
  }
}

@media @phone {
  .row-chart-section {
    margin-left: 0;
  }
}

.row-chart {
  .bg-hover,
  .visibility,
  line,
  path.domain,
  .x-axis-group {
    display: none;
  }

  .expandable {
    color: var(--pacific);
    cursor: pointer;
  }

  text {
    font-size: 16px;

    &.adjust-upwards {
      transform: translateY(-12px);
    }

    &.hidden {
      display: none;
    }
  }

  .view-more-group {
    .view-more-background {
      fill: var(--pacific-20);
    }

    .view-more-label {
      font-size: 14px;
      font-weight: bold;
      text-decoration: underline;
      fill: var(--pacific);
      text-anchor: end;
      cursor: pointer;
    }
  }
}
