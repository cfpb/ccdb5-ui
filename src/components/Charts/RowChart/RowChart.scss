@use '@cfpb/cfpb-design-system/src/abstracts' as *;
@use 'css/base' as *;

.row-chart-section {
  margin-top: $grid-gutter-width;
  margin-left: $grid-gutter-width;

  p {
    max-width: none;
  }
}

@media screen and (max-width: $layout-breakpoint-phone-max) {
  .row-chart-section {
    margin-left: 0;
  }
}

.row-chart {
  .bg-hover,
  .visibility,
  line,
  path.domain,
  .x-axis-group {
    display: none;
  }

  .expandable {
    color: var(--pacific);
    cursor: pointer;
  }

  text {
    font-size: 16px;

    &.adjust-upwards {
      transform: translateY(-12px);
    }

    &.hidden {
      display: none;
    }
  }

  .view-more-group {
    .view-more-background {
      fill: var(--pacific-20);
    }

    .view-more-label {
      font-size: 14px;
      font-weight: bold;
      text-decoration: underline;
      fill: var(--pacific);
      text-anchor: end;
      cursor: pointer;
    }
  }
}
