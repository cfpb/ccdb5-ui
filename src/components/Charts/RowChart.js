/* eslint complexity: ["error", 5] */

import './RowChart.less';
import * as d3 from 'd3';
import {
  cloneDeep,
  coalesce,
  getAllFilters,
  hashObject,
  sendAnalyticsEvent
} from '../../utils';
import { collapseRow, expandRow } from '../../actions/view';
import { miniTooltip, row } from 'britecharts';
import { changeFocus } from '../../actions/trends';
import { connect } from 'react-redux';
import { max } from 'd3-array';
import { MODE_MAP } from '../../constants';
import PropTypes from 'prop-types';
import React from 'react';
import { scrollToFocus } from '../../utils/trends';

export class RowChart extends React.Component {
  constructor( props ) {
    super( props );
    this._selectFocus = this._selectFocus.bind( this );
    this._toggleRow = this._toggleRow.bind( this );
  }

  _formatTip( value ) {
    return value.toLocaleString() + ' complaints';
  }

  _getHeight( numRows ) {
    return numRows === 1 ? 100 : numRows * 60;
  }

  _wrapText( text, width, viewMore ) {
    // ignore test coverage since this is code borrowed from d3 mbostock
    // text wrapping functions
    /* eslint-disable complexity */
    /* istanbul ignore next */
    text.each( function() {
      const innerText = d3.select( this );
      const spanWidth = viewMore ? innerText.attr( 'x' ) : 0;
      if ( innerText.node().children && innerText.node().children.length > 0 ) {
        // assuming its already split up
        return;
      }
      const words = innerText.text().split( /\s+/ ).reverse(),
        // ems
            lineHeight = 1.1,
            y = innerText.attr( 'y' ) || 0,
            dy = parseFloat( innerText.attr( 'dy' ) || 0 );

      let word,
          line = [],
          lineNumber = 0,
          wrapCount = 0,
          tspan = innerText
          .text( null )
          .append( 'tspan' )
          .attr( 'x', spanWidth )
          .attr( 'y', y )
          .attr( 'dy', dy + 'em' );

      // eslint-disable-next-line no-cond-assign
      while ( word = words.pop() ) {
        line.push( word );
        tspan.text( line.join( ' ' ) );
        if ( tspan.node().getComputedTextLength() > width ) {
          line.pop();
          tspan.text( line.join( ' ' ) );
          line = [ word ];
          tspan = innerText
            .append( 'tspan' )
            .attr( 'x', spanWidth )
            .attr( 'y', y )
            // eslint-disable-next-line no-mixed-operators
            .attr( 'dy', ++lineNumber * lineHeight + dy + 'em' )
            .text( word );
          wrapCount++;
        }
      }

      // only allow this to go through if not IE
      if ( wrapCount && !window.document.documentMode ) {
        const viewMoreBackground = d3
          .select( innerText.node().parentNode )
          .select( '.view-more-background' );
        const oldHeight = viewMoreBackground.attr( 'height' );
        // eslint-disable-next-line no-mixed-operators
        const newHeight = parseFloat( oldHeight ) + wrapCount * 12;
        viewMoreBackground.attr( 'height', newHeight );
      }
    } );
    /* eslint-enable complexity */
  }

  componentDidMount() {
    this._redrawChart();
  }

  componentDidUpdate( prevProps ) {
    const props = this.props;
    if ( hashObject( prevProps ) !== hashObject( props ) ) {
      this._redrawChart();
    }
  }

  // --------------------------------------------------------------------------
  // Event Handlers
  // eslint-disable-next-line complexity
  _redrawChart() {
    const { colorScheme, data, id, isPrintMode, total } = this.props;
    // deep copy
    // do this to prevent REDUX pollution
    const rows = cloneDeep( data ).filter( o => {
      if ( o.name && isPrintMode ) {
        // remove spacer text if we are in print mode
        return o.name.indexOf( 'Visualize trends for' ) === -1;
      }
      return true;
    } );

    if ( !rows || !rows.length || !total ) {
      return;
    }

    const tooltip = miniTooltip();
    tooltip.valueFormatter( this._formatTip );

    const ratio = total / max( rows, o => o.value );
    const chartID = '#row-chart-' + id;
    d3.selectAll( chartID + ' .row-chart' ).remove();
    const rowContainer = d3.select( chartID );

    // added padding to make up for margin
    const width = isPrintMode ?
      750 :
      rowContainer.node().getBoundingClientRect().width + 30;

    const height = this._getHeight( rows.length );
    const chart = row();
    const marginLeft = width / 4;

    // tweak to make the chart full width at desktop
    // add space at narrow width
    const marginRight = width < 600 ? 40 : -65;
    chart
      .margin( {
        left: marginLeft,
        right: marginRight,
        top: 20,
        bottom: 10
      } )
      .colorSchema( colorScheme )
      .backgroundColor( '#f7f8f9' )
      .paddingBetweenGroups( 25 )
      .enableLabels( true )
      .labelsTotalCount( total.toLocaleString() )
      .labelsNumberFormat( ',d' )
      .outerPadding( 0.1 )
      .percentageAxisToMaxRatio( ratio )
      .yAxisLineWrapLimit( 2 )
      .yAxisPaddingBetweenChart( 20 )
      .width( width )
      .wrapLabels( true )
      .height( height )
      .on( 'customMouseOver', tooltip.show )
      .on( 'customMouseMove', tooltip.update )
      .on( 'customMouseOut', tooltip.hide );

    rowContainer.datum( rows ).call( chart );
    const tooltipContainer = d3.selectAll(
      chartID + ' .row-chart .metadata-group'
    );
    tooltipContainer.datum( [] ).call( tooltip );
    this._wrapText( d3.select( chartID ).selectAll( '.tick text' ), marginLeft );

    this._wrapText(
      d3.select( chartID ).selectAll( '.view-more-label' ),
      width / 2,
      true
    );

    rowContainer.selectAll( '.y-axis-group .tick' ).on( 'click', this._toggleRow );

    rowContainer.selectAll( '.view-more-label' ).on( 'click', this._selectFocus );
  }

  _selectFocus( element ) {
    // make sure to assign a valid lens when a row is clicked
    const lens = this.props.lens === 'Overview' ? 'Product' : this.props.lens;
    const filters = coalesce( this.props.aggs, lens.toLowerCase(), [] );
    this.props.selectFocus( element, lens, filters );
  }

  _toggleRow( rowName ) {
    // fire off different action depending on if the row is expanded or not
    const { data, expandedRows } = this.props;
    const expandableRows = data.filter( o => o.isParent ).map( o => o.name );

    if ( !expandableRows.includes( rowName ) ) {
      // early exit
      return;
    }

    if ( expandedRows.includes( rowName ) ) {
      this.props.collapseRow( rowName );
    } else {
      this.props.expandRow( rowName );
    }
  }

  render() {
    return (
      this.props.total > 0 &&
        <div className="row-chart-section">
          <h3>{this.props.title}</h3>
          <p>{this.props.helperText}</p>
          <div id={'row-chart-' + this.props.id}></div>
        </div>

    );
  }
}

export const mapDispatchToProps = dispatch => ( {
  selectFocus: ( element, lens, filters ) => {
    scrollToFocus();
    let values = [];
    if ( lens === 'Company' ) {
      values.push( element.parent );
    } else {
      const filterGroup = filters.find( o => o.key === element.parent );
      const keyName = 'sub_' + lens.toLowerCase() + '.raw';
      values = filterGroup ?
        getAllFilters( element.parent, filterGroup[keyName].buckets ) :
        [];
    }
    sendAnalyticsEvent( 'Trends click', element.parent );
    dispatch( changeFocus( element.parent, lens, [ ...values ] ) );
  },
  collapseRow: rowName => {
    sendAnalyticsEvent( 'Bar chart collapsed', rowName );
    dispatch( collapseRow( rowName ) );
  },
  expandRow: rowName => {
    sendAnalyticsEvent( 'Bar chart expanded', rowName );
    dispatch( expandRow( rowName ) );
  }
} );

export const mapStateToProps = state => {
  const { tab } = state.query;
  const lens = tab === MODE_MAP ? 'Product' : state.query.lens;
  const { aggs } = state;
  const { expandedRows, isPrintMode, width } = state.view;
  return {
    aggs,
    expandedRows,
    lens,
    isPrintMode,
    tab,
    width
  };
};

export default connect( mapStateToProps, mapDispatchToProps )( RowChart );

RowChart.propTypes = {
  id: PropTypes.string.isRequired,
  colorScheme: PropTypes.oneOfType( [ PropTypes.array, PropTypes.bool ] )
    .isRequired,
  data: PropTypes.array.isRequired,
  title: PropTypes.string.isRequired,
  total: PropTypes.number.isRequired,
  isPrintMode: PropTypes.bool
};
