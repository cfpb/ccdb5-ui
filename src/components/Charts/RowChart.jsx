/* eslint complexity: ["error", 5] */

import './RowChart.less'
import * as d3 from 'd3'
import { changeFocus, collapseTrend, expandTrend } from '../../actions/trends'
import { cloneDeep, coalesce, getAllFilters, hashObject } from '../../utils'
import { miniTooltip, row } from 'britecharts'
import { connect } from 'react-redux'
import { max } from 'd3-array'
import { MODE_MAP } from '../../constants'
import PropTypes from 'prop-types'
import React from 'react'
import { scrollToFocus } from '../../utils/trends'

export class RowChart extends React.Component {
  constructor( props ) {
    super( props )
    this._selectFocus = this._selectFocus.bind( this )
    this._toggleRow = this._toggleRow.bind( this )
  }

  _formatTip( value ) {
    return value.toLocaleString() + ' complaints'
  }

  _getHeight( numRows ) {
    return numRows === 1 ? 100 : numRows * 60
  }

  _wrapText( text, width ) {
    // ignore test coverage since this is code borrowed from d3 mbostock
    // text wrapping functions

    /* eslint-disable complexity */
    /* istanbul ignore next */
    text.each( function() {
      const innerText = d3.select( this )
      if ( innerText.node().children && innerText.node().children.length > 0 ) {
        // assuming its already split up
        return
      }
      const words = innerText.text().split( /\s+/ ).reverse(),
            // ems
            lineHeight = 1.1,
            y = innerText.attr( 'y' ) || 0,
            dy = parseFloat( innerText.attr( 'dy' ) || 0 )

      let word,
          line = [],
          lineNumber = 0,
          tspan = innerText.text( null ).append( 'tspan' ).attr( 'x', 0 )
            .attr( 'y', y ).attr( 'dy', dy + 'em' )

      // eslint-disable-next-line no-cond-assign
      while ( word = words.pop() ) {
        line.push( word )
        tspan.text( line.join( ' ' ) )
        if ( tspan.node().getComputedTextLength() > width ) {
          line.pop()
          tspan.text( line.join( ' ' ) )
          line = [ word ]
          tspan = innerText.append( 'tspan' )
            .attr( 'x', 0 )
            .attr( 'y', y )
            // eslint-disable-next-line no-mixed-operators
            .attr( 'dy', ++lineNumber * lineHeight + dy + 'em' )
            .text( word )
        }
      }
    } )
    /* eslint-enable complexity */

  }

  componentDidMount() {
    this._redrawChart()
  }

  componentDidUpdate( prevProps ) {
    const props = this.props
    if ( hashObject( prevProps ) !== hashObject( props ) ) {
      this._redrawChart()
    }
  }

  // --------------------------------------------------------------------------
  // Event Handlers
  // eslint-disable-next-line complexity
  _redrawChart() {
    const {
      colorScheme, data, id, printMode, total
    } = this.props
    // deep copy
    // do this to prevent REDUX pollution
    const rows = cloneDeep( data ).filter( o => {
      return o.name ? o.name.indexOf( 'Visualize trends for' ) === -1 : true
    } )

    if ( !rows || !rows.length || !total ) {
      return
    }

    const tooltip = miniTooltip()
    tooltip.valueFormatter( this._formatTip )

    const ratio = total / max( rows, o => o.value )
    const chartID = '#row-chart-' + id
    d3.selectAll( chartID + ' .row-chart' ).remove()
    const rowContainer = d3.select( chartID )
    const width = printMode ? 750 :
      rowContainer.node().getBoundingClientRect().width
    const height = this._getHeight( rows.length )
    const chart = row()
    const marginLeft = width / 4

    // tweak to make the chart full width at desktop
    // add space at narrow width
    const marginRight = width < 600 ? 20 : -80
    chart.margin( {
      left: marginLeft,
      right: marginRight,
      top: 20,
      bottom: 10
    } )
      .colorSchema( colorScheme )
      .backgroundColor( '#f7f8f9' )
      .paddingBetweenGroups( 25 )
      .enableLabels( true )
      .labelsTotalCount( total.toLocaleString() )
      .labelsNumberFormat( ',d' )
      .outerPadding( 0.1 )
      .percentageAxisToMaxRatio( ratio )
      .yAxisLineWrapLimit( 2 )
      .yAxisPaddingBetweenChart( 20 )
      .width( width )
      .wrapLabels( true )
      .height( height )
      .on( 'customMouseOver', tooltip.show )
      .on( 'customMouseMove', tooltip.update )
      .on( 'customMouseOut', tooltip.hide )

    rowContainer.datum( rows ).call( chart )
    const tooltipContainer =
      d3.selectAll( chartID + ' .row-chart .metadata-group' )
    tooltipContainer.datum( [] ).call( tooltip )
    this._wrapText( d3.select( chartID ).selectAll( '.tick text' ), marginLeft )

    this._wrapText( d3.select( chartID )
      .selectAll( '.view-more-label' ), width / 2 )

    rowContainer
      .selectAll( '.y-axis-group .tick' )
      .on( 'click', this._toggleRow )

    rowContainer
      .selectAll( '.view-more-label' )
      .on( 'click', this._selectFocus )

  }

  _selectFocus( element ) {
    // make sure to assign a valid lens when a row is clicked
    const lens = this.props.lens === 'Overview' ? 'Product' : this.props.lens
    const filters = coalesce( this.props.aggs, lens.toLowerCase(), [] )
    this.props.selectFocus( element, lens, filters )
  }

  _toggleRow( rowName ) {
    // fire off different action depending on if the row is expanded or not
    const { expandableRows, expandedTrends } = this.props
    if ( expandableRows.includes( rowName ) ) {
      if ( expandedTrends.includes( rowName ) ) {
        this.props.collapseRow( rowName )
      } else {
        this.props.expandRow( rowName )
      }
    }
  }


  render() {
    return (
      this.props.total > 0 &&
      <div className="row-chart-section">
        <h3>{ this.props.title }</h3>
        <p>{ this.props.helperText }</p>
        <div id={ 'row-chart-' + this.props.id }>
        </div>
      </div>
    )
  }
}

export const mapDispatchToProps = dispatch => ( {
  selectFocus: ( element, lens, filters ) => {
    scrollToFocus()
    let values = []
    if ( lens === 'Company' ) {
      values.push( element.parent )
    } else {
      const filterGroup = filters.find( o => o.key === element.parent )
      const keyName = 'sub_' + lens.toLowerCase() + '.raw'
      values = filterGroup ?
        getAllFilters( element.parent, filterGroup[keyName].buckets ) : []
    }

    dispatch( changeFocus( element.parent, lens, [ ...values ] ) )
  },
  collapseRow: rowName => {
    dispatch( collapseTrend( rowName.trim() ) )
  },
  expandRow: rowName => {
    dispatch( expandTrend( rowName.trim() ) )
  }
} )

export const mapStateToProps = state => {
  const { tab } = state.query
  const lens = tab === MODE_MAP ? 'Product' : state.query.lens
  const { aggs } = state
  const { expandableRows, expandedTrends } =
    coalesce( state, tab.toLowerCase(), {} )
  const { printMode, width } = state.view
  return {
    aggs,
    expandableRows,
    expandedTrends,
    lens,
    printMode,
    tab,
    width
  }
}

RowChart.propTypes = {
  id: PropTypes.string.isRequired,
  colorScheme: PropTypes.oneOfType( [
    PropTypes.array,
    PropTypes.bool
  ] ).isRequired,
  data: PropTypes.array.isRequired,
  title: PropTypes.string.isRequired,
  total: PropTypes.number.isRequired
}

export default connect( mapStateToProps, mapDispatchToProps )( RowChart )
