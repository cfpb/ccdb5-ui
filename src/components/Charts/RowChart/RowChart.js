import './RowChart.scss';
import * as d3 from 'd3';
import { max } from 'd3-array';
import { miniTooltip, row } from 'britecharts';
import PropTypes from 'prop-types';
import { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { scrollToFocus } from '../../../utils/trends';
import { focusChanged } from '../../../reducers/trends/trendsSlice';
import { rowCollapsed, rowExpanded } from '../../../reducers/view/viewSlice';
import { selectTrendsLens } from '../../../reducers/trends/selectors';
import {
  selectViewIsPrintMode,
  selectViewExpandedRows,
  selectViewTab,
  selectViewWidth,
} from '../../../reducers/view/selectors';
import { coalesce, getAllFilters, sendAnalyticsEvent } from '../../../utils';
import { MODE_MAP } from '../../../constants';
import { useGetAggregations } from '../../../api/hooks/useGetAggregations';

export const RowChart = ({
  helperText,
  id,
  colorScheme,
  data,
  title,
  total,
}) => {
  const dispatch = useDispatch();
  const { data: aggs } = useGetAggregations();
  const tab = useSelector(selectViewTab);
  const trendsLens = useSelector(selectTrendsLens);
  const expandedRows = useSelector(selectViewExpandedRows);
  const isPrintMode = useSelector(selectViewIsPrintMode);
  const width = useSelector(selectViewWidth);
  const lens = tab === MODE_MAP ? 'Product' : trendsLens;

  useEffect(() => {
    const chartID = '#row-chart-' + id;

    const formatTip = (value) => {
      return value.toLocaleString() + ' complaints';
    };

    const wrapText = (text, width, viewMore) => {
      // ignore test coverage since this is code borrowed from d3 mbostock
      // text wrapping functions

      /* istanbul ignore next */
      text.each(function () {
        const innerText = d3.select(this);
        const spanWidth = viewMore ? innerText.attr('x') : 0;
        if (innerText.node().children && innerText.node().children.length > 0) {
          // assuming its already split up
          return;
        }
        const words = innerText.text().split(/\s+/).reverse(),
          // ems
          lineHeight = 1.1,
          // eslint-disable-next-line id-length
          y = innerText.attr('y') || 0,
          dy = parseFloat(innerText.attr('dy') || 0);

        let word,
          line = [],
          lineNumber = 0,
          wrapCount = 0,
          tspan = innerText
            .text(null)
            .append('tspan')
            .attr('x', spanWidth)
            .attr('y', y)
            .attr('dy', dy + 'em');

        while ((word = words.pop())) {
          line.push(word);
          tspan.text(line.join(' '));
          if (tspan.node().getComputedTextLength() > width) {
            line.pop();
            tspan.text(line.join(' '));
            line = [word];
            tspan = innerText
              .append('tspan')
              .attr('x', spanWidth)
              .attr('y', y)

              .attr('dy', ++lineNumber * lineHeight + dy + 'em')
              .text(word);
            wrapCount++;
          }
        }

        if (wrapCount) {
          const viewMoreBackground = d3
            .select(innerText.node().parentNode)
            .select('.view-more-background');
          const oldHeight = viewMoreBackground.attr('height');

          const newHeight = parseFloat(oldHeight) + wrapCount * 12;
          viewMoreBackground.attr('height', newHeight);
        }
      });
    };

    const collapseARow = (rowName) => {
      sendAnalyticsEvent('Bar chart collapsed', rowName);
      dispatch(rowCollapsed(rowName));
    };

    const expandARow = (rowName) => {
      sendAnalyticsEvent('Bar chart expanded', rowName);
      dispatch(rowExpanded(rowName));
    };

    const selectFocus = (element) => {
      const focusName = element.target.__data__;
      // make sure to assign a valid lens when a row is clicked
      const aLens = lens === 'Overview' ? 'Product' : lens;
      const filters = coalesce(aggs, aLens.toLowerCase(), []);
      scrollToFocus();
      let values = [];
      if (lens === 'Company') {
        values.push(focusName.parent);
      } else {
        const filterGroup = filters.find((obj) => obj.key === focusName.parent);
        const keyName = 'sub_' + lens.toLowerCase() + '.raw';
        values = filterGroup
          ? getAllFilters(focusName.parent, filterGroup[keyName].buckets)
          : [];
      }
      sendAnalyticsEvent('Trends click', focusName.parent);
      dispatch(focusChanged(focusName.parent, lens, [...values]));
    };

    const toggleRow = (element) => {
      const rowName = element.target.__data__;
      // fire off different action depending on if the row is expanded or not
      const expandableRows = data
        .filter((obj) => obj.isParent)
        .map((obj) => obj.name);

      if (!expandableRows.includes(rowName)) {
        // early exit
        return;
      }

      if (expandedRows.includes(rowName)) {
        collapseARow(rowName);
      } else {
        expandARow(rowName);
      }
    };

    if (!data) {
      return;
    }

    // do this to prevent REDUX pollution
    const rows = data.filter((obj) => {
      if (obj.name && isPrintMode) {
        // remove spacer text if we are in print mode
        return obj.name.indexOf('Visualize trends for') === -1;
      }
      return true;
    });

    if (!rows || !rows.length || !total) {
      return;
    }

    const tooltip = miniTooltip();
    tooltip.valueFormatter(formatTip);

    const ratio = total / max(rows, (obj) => obj.value);
    const rowContainer = d3.select(chartID);

    // added padding to make up for margin
    const containerWidth = isPrintMode
      ? 750
      : rowContainer.node().getBoundingClientRect().width + 30;

    const height = rows.length === 1 ? 100 : rows.length * 60;
    const chart = row();
    const marginLeft = containerWidth / 4;

    // tweak to make the chart full width at desktop
    // add space at narrow width
    const marginRight = containerWidth < 600 ? 40 : -65;

    chart
      .margin({
        left: marginLeft,
        right: marginRight,
        top: 20,
        bottom: 10,
      })
      .colorSchema(colorScheme)
      .backgroundColor('#f7f8f9')
      .paddingBetweenGroups(25)
      .enableLabels(true)
      .labelsTotalCount(total.toLocaleString())
      .labelsNumberFormat(',d')
      .outerPadding(0.1)
      .percentageAxisToMaxRatio(ratio)
      .yAxisLineWrapLimit(2)
      .yAxisPaddingBetweenChart(20)
      .width(containerWidth)
      .wrapLabels(true)
      .height(height)
      .on('customMouseOver', tooltip.show)
      .on('customMouseMove', tooltip.update)
      .on('customMouseOut', tooltip.hide);

    rowContainer.datum(rows).call(chart);
    const tooltipContainer = d3.selectAll(
      chartID + ' .row-chart .metadata-group',
    );
    tooltipContainer.datum([]).call(tooltip);

    wrapText(d3.select(chartID).selectAll('.tick text'), marginLeft);
    wrapText(d3.select(chartID).selectAll('.view-more-label'), width / 2, true);

    rowContainer.selectAll('.y-axis-group .tick').on('click', toggleRow);
    rowContainer.selectAll('.view-more-label').on('click', selectFocus);

    return () => {
      d3.selectAll(chartID + ' .row-chart').remove();
    };
  }, [
    dispatch,
    aggs,
    colorScheme,
    data,
    expandedRows,
    id,
    isPrintMode,
    lens,
    total,
    width,
  ]);

  if (!data) {
    return null;
  }

  return total ? (
    <div className="row-chart-section">
      <h3>{title}</h3>
      <p>{helperText}</p>
      <div id={'row-chart-' + id} data-testid={'row-chart-' + id} />
    </div>
  ) : null;
};

RowChart.propTypes = {
  helperText: PropTypes.string,
  id: PropTypes.string.isRequired,
  colorScheme: PropTypes.oneOfType([PropTypes.array, PropTypes.bool])
    .isRequired,
  data: PropTypes.array.isRequired,
  title: PropTypes.string.isRequired,
  total: PropTypes.number,
};
