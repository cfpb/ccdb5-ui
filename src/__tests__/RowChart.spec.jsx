import configureMockStore from 'redux-mock-store'
import { mapStateToProps, RowChart } from '../RowChart'
import { mount, shallow } from 'enzyme'
import { Provider } from 'react-redux'
import React from 'react'
import renderer from 'react-test-renderer'
import thunk from 'redux-thunk'

// this is how you override and mock an imported constructor
jest.mock( 'britecharts', () => {
  const props = [
    'row', 'margin', 'backgroundColor', 'enableLabels', 'labelsSize',
    'labelsTotalCount', 'labelsNumberFormat', 'outerPadding',
    'percentageAxisToMaxRatio', 'yAxisLineWrapLimit',
    'yAxisPaddingBetweenChart', 'width', 'wrapLabels', 'height'
  ]

  const mock = {}

  for ( let i = 0; i < props.length; i++ ) {
    const propName = props[i]
    mock[propName] = jest.fn().mockImplementation( () => {
      return mock
    } )
  }

  return mock
} )

jest.mock( 'd3', () => {
  const props = [
    'select', 'each', 'node', 'getBoundingClientRect', 'width', 'datum', 'call',
    'remove', 'selectAll'
  ]

  const mock = {}

  for ( let i = 0; i < props.length; i++ ) {
    const propName = props[i]
    mock[propName] = jest.fn().mockImplementation( () => {
      return mock
    } )
  }

  return mock
} )

function setupSnapshot() {
  const middlewares = [ thunk ]
  const mockStore = configureMockStore( middlewares )
  const store = mockStore( {
    map: {}
  } )

  return renderer.create(
    <Provider store={ store }>
      <RowChart aggtype={'foo'}/>
    </Provider>
  )
}

describe( 'component: RowChart', () => {
  describe( 'initial state', () => {
    it( 'renders without crashing', () => {
      const target = setupSnapshot()
      let tree = target.toJSON()
      expect( tree ).toMatchSnapshot()
    } )
  } )

  describe( 'componentDidUpdate', () => {
    let mapDiv

    beforeEach( () => {
      mapDiv = document.createElement( 'div' )
      mapDiv.setAttribute( 'id', 'row-chart-foo' )
      window.domNode = mapDiv
      document.body.appendChild( mapDiv )
    } )

    afterEach( () => {
      const div = document.getElementById( 'row-chart-foo' )
      if ( div ) {
        document.body.removeChild( div )
      }
      jest.clearAllMocks()
    } )

    it( 'does nothing when no data', () => {
      const target = shallow( <RowChart data={ [] } aggtype={'foo'}/> )
      target._redrawChart = jest.fn()
      target.setProps( { data: [] } )
      expect(  target._redrawChart ).toHaveBeenCalledTimes( 0 )
    } )

    it( 'trigger a new update when data changes', () => {
      const target = shallow( <RowChart data={ [ 23, 4, 3 ] } aggtype={'foo'} total={1000}/> )
      target._redrawChart = jest.fn()
      const sp = jest.spyOn(target.instance(), '_redrawChart')
      target.setProps( { data: [ 2, 5 ] } )
      expect( sp ).toHaveBeenCalledTimes( 1 )
    } )
  } )

  describe( 'mapStateToProps', () => {
    it( 'maps state and props', () => {
      const state = {
        query: {
          baz: [ 1, 2, 3 ]
        },
        map: {
          baz: [ 1, 2, 3 ]
        },
        results: {
          total: 100
        }
      }
      const ownProps = {
        aggtype: 'baz'
      }
      let actual = mapStateToProps( state, ownProps )
      expect( actual ).toEqual( {
        data: [],
        total: 100
      } )
    } )
  } )

  describe('helper functions', ()=>{
    it('gets height based on number of rows', ()=>{
      const target = mount(<RowChart />)
      let res = target.instance()._getHeight(1)
      expect(res).toEqual(100)
      res = target.instance()._getHeight(5)
      expect(res).toEqual(300)
    })
  })

} )
