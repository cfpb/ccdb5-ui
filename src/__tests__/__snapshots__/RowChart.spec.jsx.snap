// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: RowChart initial state renders without crashing 1`] = `
<div>
  <h3>
    foo
     by complaint volume (5 highest counts)
  </h3>
  <div
    id="row-chart-foo"
  />
</div>
`;
